test/test_charm.py

```diff
@@ -15,9 +17,14 @@ class TestCharm(unittest.TestCase):
     def setUp(self):
         self.tmpdir = Path(tempfile.mkdtemp())
 
+        # Avoid introducing side-effects for other tests by using a separate type.
+        CharmBase.on = types.new_class('TestCharmEvents', bases=(CharmEvents,))()
```

Comment:
This can be normal Python code:
```
class TestCharmEvents(CharmEvents): pass
```

Reply:
I originally suggested `new_class` because I assumed that the `TestCharmEvents` class definition would be persistent even inside a function call.  But I tested it and realized that assumption was wrong, which makes sense now that I think about it again.  Good to know, and this is much more simple.